Add OTEL_EXPORTER environment variable

trask opened the pull request:
Fixes #710

## Changes

Adds `OTEL_EXPORTER` environment variable to select the exporter to be used.

Adds `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable to configure the ingest endpoint for both OTLP spans and metrics.

## Notes

1. I added `OTEL_EXPORTER_OTLP_ENDPOINT` in this PR because I think it makes the default value `OTEL_EXPORTER=otlp` make more sense.

   (It also happens to be the env var that the Java SDK uses currently, and seems simpler for most OTLP users, who will be sending spans and metrics to the same endpoint)

   An alternative would be to specify the default value as `OTEL_EXPORTER=otlp_span,otlp_metric`.

2. I went with the singular `OTEL_EXPORTER` over the plural `OTEL_EXPORTERS`, since the common use case is to configure just one exporter and so this seems less confusing.

anuraaga commented on specification/sdk-environment-variables.md:
```diff
@@ -49,6 +55,12 @@ The goal of this specification is to unify the environment variable names betwee
+
+| Name          | Description                                                                  | Default | Notes                                                              |
+| ------------- | ---------------------------------------------------------------------------- | ------- | ------------------------------------------------------------------ |
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  | Other known values: "jaeger", "zipkin", "otlp_span", "otlp_metric" |
```
prometheus?

carlosalberto replied:
Probably move the list of known values outside the table? Could be handy in case we add more ;)

trask replied:
done

iNikem commented on specification/sdk-environment-variables.md:
```diff
@@ -49,6 +55,12 @@ The goal of this specification is to unify the environment variable names betwee
+
+| Name          | Description                                                                  | Default | Notes                                                              |
+| ------------- | ---------------------------------------------------------------------------- | ------- | ------------------------------------------------------------------ |
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  | Other known values: "jaeger", "zipkin", "otlp_span", "otlp_metric" |
```
How can I specify different exporters for spans and metrics? E.g. OTLP for spans, but Prometheus for metrics?

codeboten replied:
Would it make sense to support `OTEL_EXPORTER_SPAN` `OTEL_EXPORTER_METRIC` for differentiating between the two here with the default for both to be `otlp`?

trask replied:
@iNikem `OTEL_EXPORTER=otlp_span,prometheus`

@codeboten see comment below: https://github.com/open-telemetry/opentelemetry-specification/pull/943#issuecomment-692449625

iNikem commented on specification/sdk-environment-variables.md:
```diff
@@ -19,19 +19,25 @@ The goal of this specification is to unify the environment variable names betwee
+| Name                             | Description                                | Default | Notes                                              |
+| -------------------------------- | ------------------------------------------ | ------- | -------------------------------------------------- |
+| OTEL_EXPORTER_OTLP_SPAN_TIMEOUT  | Max waiting time to export each span batch | -       |                                                    |
+| OTEL_EXPORTER_OTLP_SPAN_ENDPOINT | Ingest endpoint for OTLP spans             | -       | Overrides OTEL_EXPORTER_OTLP_ENDPOINT if non-empty |
```
Should default value be localhost and default OTLP port in the collector?

trask replied:
now pointing this section to https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md which has these default values 👍

codeboten commented on specification/sdk-environment-variables.md:
```diff
@@ -19,19 +19,25 @@ The goal of this specification is to unify the environment variable names betwee
+
+| Name                        | Description                                     | Default |
+| --------------------------- | ----------------------------------------------- | ------- |
+| OTEL_EXPORTER_OTLP_ENDPOINT | Ingest endpoint for both OTLP spans and metrics | -       |
```
It's worth noting that there are additional configuration options defined here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md

trask replied:
thx, I hadn't seen that!

codeboten reviewed (commented):
Is there a mention in the spec as to what the known values that should be used to identify exporters are and how exporters should register these values? That would be useful for anyone implementing additional exporters.

SergeyKanzhelev commented:
> Adds `OTEL_EXPORTER` environment variable to select the exporter to be used.

would it make more sense for every exporter to keep `_ENABLED` variable (like `OTEL_EXPORTER_OTLP_SPAN_ENABELD` and `OTEL_EXPORTER_JAEGER_ENABLED`)? This way we can enable any number of exporters and no need to keep the centralized list. Also it is more extensible as the list of exporters can be extended in very logical way.

Having `OTEL_EXPORTER` is more aligned with the existence of a single batch processor setting. But I'd argue it's more logical to have enable flag configured with the exporter.

trask commented:
There are two alternatives suggested above. I like both of them, except around handling of the default value.

1. Suggestion: split into two env vars, one for span exporter (`OTEL_EXPORTER_SPAN`) and one for metric exporter (`OTEL_EXPORTER_METRIC`)

   The default in this case would be `OTEL_EXPORTER_SPAN=otlp` and `OTEL_EXPORTER_METRIC=otlp`.

   If someone wants to use jaeger, would they need to set both `OTEL_EXPORTER_SPAN=jaeger` and `OTEL_EXPORTER_METRIC=none`?

2. Suggestion: split out into a boolean flag for each exporter (`OTEL_EXPORTER_*_ENABLED`)

   The default in this case would be `OTEL_EXPORTER_OTLP_ENABLED=true` and all others `OTEL_EXPORTER_*_ENABLED=false`.

   If someone wants to use jaeger, would they need to set both `OTEL_EXPORTER_JAEGER_ENABLED=true` and `OTEL_EXPORTER_OTLP_ENABLED=false`.

Wondering if anyone has suggestions how to handle default value better in either of these cases? Or should we stick with the single environment variable `OTEL_EXPORTER` with default value `otlp`?

iNikem commented on specification/sdk-environment-variables.md:
```diff
@@ -49,6 +39,14 @@ The goal of this specification is to unify the environment variable names betwee
+| ------------- | ---------------------------------------------------------------------------- | ------- |
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  |
+
+Known values for OTEL_EXPORTER are: "otlp", "jaeger", "zipkin", "prometheus", "otlp_span", "otlp_metric".
```
Probably should explicitly say that `otlp` means `otlp_span+otlp_metric`?

trask replied:
done

owais replied:
What about 3rd party or contrib exporters? This is language specific but spec should probably mention if we want to support that. 

For example, Python could support something like `OTEL_EXPORTER=otlp_span,my_exporter.MetricExporter'. Any values present in the list other than the well-known ones mentioned in the spec will have to be fully-qualified import paths to exporter implementations. Another issue this brings up is how do we tell if a custom exporter exports metrics, spans or logs. 

1. Each implementation could use some sort of type-checking to verify what kind of exporter something implements. If something implements multiple exporters, do we instantiate a single instance for all supported types or multiple instances per type?

2. Go with the other suggestion to have one env var per type i.e, `OTEL_SPAN_EXPORTER`, `OTEL_METRIC_EXPORTER` etc. 

Java today supports using an external exporter  by pointing to a .jar file `-Dotel.exporter.jar=path/to/external-exporter.jar`.  https://github.com/open-telemetry/opentelemetry-java-instrumentation#getting-started

I also have an open PR for Python that supports specifying import paths to load custom exporters.

iNikem commented:
I prefer the third alternative, which is currently written in PR :)

trask commented:
In case we go with `OTEL_EXPORTER`, and because I ❤️ naming, does anyone care between:
* `OTEL_EXPORTER=otlp_span`
* `OTEL_EXPORTER=otlp-span`
* `OTEL_EXPORTER=otlpspan`

Exporter names seem sort of like propagator names, not sure if there's any convention we can follow from those?

iNikem commented:
For some reason `OTEL_EXPORTER=otlp_span` seems the most natural to me

SergeyKanzhelev commented:
@trask you got me on default values. Single variable is better from this perspective.

As for naming, I again tend to lean to separate variables for metrics and spans. It feels like two separate pieces of configuration semantically.

Since I also ❤️  naming, should we get PM to this PR to help nail the desired semantic of this config settings ;-)?

codeboten reviewed (approved):
Thanks for the updates, this LGTM

trask commented:
> (edit: I assumed "because I ❤️ naming" was a sarcasm. If not, than mine is =))

Mine was intentionally ambiguous 😂. I have a complicated relationship with naming.

tigrannajaryan reviewed (approved):
LGTM

SergeyKanzhelev commented:
Last edit was a day ago, so plan to merge tomorrow morning unless somebody will comment with blocking concerns

owais commented:
Other than OTLP, most other exporters expect an explicit value for "service name". OTLP expects a `service_name` resource attribute. The spec already specifies a way to configure resource attributes globally for a service using environment variables (`OTEL_RESOURCE_ATTRIBUTES`) so OTLP is covered but this leaves other exporters without a way to set the service name. We can either add a common `OTEL_SERVICE_NAME` env var or one env var per exporter like `OTEL_EXPORTER_JAEGER_SERVICE_NAME`.

owais commented on specification/sdk-environment-variables.md:
```diff
@@ -49,6 +39,16 @@ The goal of this specification is to unify the environment variable names betwee
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  |
+
+Known values for OTEL_EXPORTER are: "otlp", "jaeger", "zipkin", "prometheus", "otlp_span", "otlp_metric".
+
```
What about 3rd party or contrib exporters? This is language specific but spec should probably mention if we want to support that.

For example, Python could support something like `OTEL_EXPORTER=otlp_span,my_exporter.MetricExporter'. Any values present in the list other than the well-known ones mentioned in the spec will have to be fully-qualified import paths to exporter implementations. Another issue this brings up is how do we tell if a custom exporter exports metrics, spans or logs.

    Each implementation could use some sort of type-checking to verify what kind of exporter something implements. If something implements multiple exporters, do we instantiate a single instance for all supported types or multiple instances per type?

    Go with the other suggestion to have one env var per type i.e, OTEL_SPAN_EXPORTER, OTEL_METRIC_EXPORTER etc.

Java today supports using an external exporter by pointing to a .jar file -Dotel.exporter.jar=path/to/external-exporter.jar. https://github.com/open-telemetry/opentelemetry-java-instrumentation#getting-started

I also have an open PR for Python that supports specifying import paths to load custom exporters.

trask replied:
I'd like to suggest that custom exporters are out of scope for this PR, can you file a separate issue/PR for this?

owais replied:
What this PR suggest does not close the door for supporting custom exporters later so dealing with that in a separate PR sounds good to me.

trask commented:
> We can either add a common `OTEL_SERVICE_NAME` env var or one env var per exporter like `OTEL_EXPORTER_JAEGER_SERVICE_NAME`.

Hey @owais, check out (and vote for!) #709
